From 76b1dcf494ada0f7245eac99f7e5c45216aa1f5b Mon Sep 17 00:00:00 2001
From: Ramiro Polla <ramiro.polla@gmail.com>
Date: Thu, 27 Jun 2013 01:59:38 -0300
Subject: [PATCH 5/6] aircrack-ptw-lib: split correct() to simplify future
 optimizations

---
 src/aircrack-ptw-lib.c |   29 ++++++++++++++++++-----------
 1 file changed, 18 insertions(+), 11 deletions(-)

diff --git a/src/aircrack-ptw-lib.c b/src/aircrack-ptw-lib.c
index 3b612d5..5962d2b 100644
--- a/src/aircrack-ptw-lib.c
+++ b/src/aircrack-ptw-lib.c
@@ -175,6 +175,22 @@ static uint8_t rc4update(rc4state * state) {
 	return state->s[k];
 }
 
+static int rc4test(uint8_t *key, int keylen, uint8_t *iv, uint8_t *keystream)
+{
+	uint8_t keybuf[PTW_KSBYTES];
+	rc4state rc4state;
+	int j;
+	memcpy(&keybuf[IVBYTES], key, keylen);
+	memcpy(keybuf, iv, IVBYTES);
+	rc4init(keybuf, keylen+IVBYTES, &rc4state);
+	for (j = 0; j < TESTBYTES; j++) {
+		if  ((rc4update(&rc4state) ^ keystream[j]) != 0) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
 // For sorting
 static int comparesorthelper(const void * ina, const void * inb) {
 	sorthelper * a = (sorthelper * ) ina;
@@ -224,10 +240,7 @@ static void guesskeybytes(int ivlen, uint8_t * iv, uint8_t * keystream, uint8_t
  */
 static int correct(PTW_attackstate * state, uint8_t * key, int keylen) {
 	int i;
-	int j;
 	int k;
-	uint8_t keybuf[PTW_KSBYTES];
-	rc4state rc4state;
 
 	// We need at least 3 sessions to be somehow certain
 	if (state->sessions_collected < 3) {
@@ -238,14 +251,8 @@ static int correct(PTW_attackstate * state, uint8_t * key, int keylen) {
 
 	k = rand()%(state->sessions_collected-10);
 	for ( i=k; i < k+10; i++) {
-		memcpy(&keybuf[IVBYTES], key, keylen);
-		memcpy(keybuf, state->sessions[i].iv, IVBYTES);
-		rc4init(keybuf, keylen+IVBYTES, &rc4state);
-		for (j = 0; j < TESTBYTES; j++) {
-			if  ((rc4update(&rc4state) ^ state->sessions[i].keystream[j]) != 0) {
-				return 0;
-			}
-		}
+		if (!rc4test(key, keylen, state->sessions[i].iv, state->sessions[i].keystream))
+			return 0;
 	}
 	return 1;
 }
-- 
1.7.9.5

