From f28ee0f98336b9f63923c1aec974f27c301ea78a Mon Sep 17 00:00:00 2001
From: Ramiro Polla <ramiro.polla@gmail.com>
Date: Sun, 23 Jun 2013 13:38:14 -0300
Subject: [PATCH 2/6] aircrack-ptw-lib: optimize integer sort helpers

---
 src/aircrack-ptw-lib.c |   16 ++--------------
 1 file changed, 2 insertions(+), 14 deletions(-)

diff --git a/src/aircrack-ptw-lib.c b/src/aircrack-ptw-lib.c
index 55af552..0025f71 100644
--- a/src/aircrack-ptw-lib.c
+++ b/src/aircrack-ptw-lib.c
@@ -119,13 +119,7 @@ PTW_tableentry keytable[KEYHSBYTES][n];
 static int compare(const void * ina, const void * inb) {
 	PTW_tableentry * a = (PTW_tableentry * )ina;
 	PTW_tableentry * b = (PTW_tableentry * )inb;
-	if (a->votes > b->votes) {
-		return -1;
-	} else if (a->votes == b->votes) {
-		return 0;
-	} else {
-		return 1;
-	}
+	return b->votes - a->votes;
 }
 
 // For sorting
@@ -182,13 +176,7 @@ static uint8_t rc4update(rc4state * state) {
 static int comparesorthelper(const void * ina, const void * inb) {
 	sorthelper * a = (sorthelper * ) ina;
 	sorthelper * b = (sorthelper * ) inb;
-	if (a->distance > b->distance) {
-		return 1;
-	} else if (a->distance == b->distance) {
-		return 0;
-	} else {
-		return -1;
-	}
+	return a->distance - b->distance;
 }
 
 /*
-- 
1.7.9.5

