Index: scripts/airoscript-ng/Makefile
===================================================================
--- scripts/airoscript-ng/Makefile	(revisione 2449)
+++ scripts/airoscript-ng/Makefile	(copia locale)
@@ -25,7 +25,6 @@
 	 install-locale \
 	 install-desktop
 
-
 installdirs:
 	@$(SHELL) ./.mkinstalldirs $(bindir) $(datadir) \
         			$(docdir)/html/images $(etcdir) \
@@ -99,7 +98,7 @@
 	@for i in $(docdir)/*.1 ; do rm -rf $(mandir)/man1/$$i ; done
 	@for i in $(srcdir)/locale/* ; do \
 		if [ -d $$i ] && [ $$i != "." ] && [ $$i != ".." ]; then \
-			make -s -C $$i uninstall localedir=$(locale) ; \
+			make -s -C $$i uninstall localedir="$(locale)" ; \
 		fi ; \
 	done
 	@echo "Uninstalled succesfully"
