Index: Makefile
===================================================================
--- Makefile	(revisione 2266)
+++ Makefile	(copia locale)
@@ -96,7 +96,7 @@
 	@rm -r $(datadir)
 	@rm -r $(docdir)
 	@rm $(mandir)/man1/airoscript-ng.1
-	@rm -f $(picdir)/airoscript-ng.png $(picdir)/airoscript-ng.desktop $(picdir)/airoscript-ng_gtk.desktop
+	@rm -f $(picdir)/airoscript-ng.png $(appdir)/airoscript-ng.desktop $(appdir)/airoscript-ng_gtk.desktop
 	@for i in $(docdir)/*.1 ; do rm -rf $(mandir)/man1/$$i ; done
 	@for i in $(srcdir)/locale/* ; do \
 		if [ -d $$i ] && [ $$i != "." ] && [ $$i != ".." ]; then \
