Index: aireplay-ng.c
===================================================================
--- aireplay-ng.c	(revision 2312)
+++ aireplay-ng.c	(working copy)
@@ -2862,6 +2862,7 @@
             if( read( dev.fd_rtc, &n, sizeof( n ) ) < 0 )
             {
                 perror( "read(/dev/rtc) failed" );
+                free(arp);
                 return( 1 );
             }
 
@@ -2904,14 +2905,18 @@
                     ticks[0] = 0;
 
                 if( send_packet( arp[arp_off1].buf,
-                                 arp[arp_off1].len ) < 0 )
+                                 arp[arp_off1].len ) < 0 ){
+                    free(arp);
                     return( 1 );
+                }
 
                 if( ((double)ticks[0]/(double)RTC_RESOLUTION)*(double)opt.r_nbpps > (double)nb_pkt_sent  )
                 {
                     if( send_packet( arp[arp_off1].buf,
-                                    arp[arp_off1].len ) < 0 )
+                                    arp[arp_off1].len ) < 0 ){
+                        free(arp);
                         return( 1 );
+                    }
                 }
 
                 if( ++arp_off1 >= nb_arp )
@@ -2927,7 +2932,10 @@
 
             caplen = read_packet( h80211, sizeof( h80211 ), NULL );
 
-            if( caplen  < 0 ) return( 1 );
+            if( caplen  < 0 ){
+                free(arp);
+                return( 1 );
+            }
             if( caplen == 0 ) continue;
         }
         else
@@ -3130,6 +3138,7 @@
 
                 if( ( arp[nb_arp].buf = malloc( 128 ) ) == NULL ) {
                     perror( "malloc failed" );
+                    free(arp);
                     return( 1 );
                 }
 
@@ -3158,6 +3167,7 @@
 
                 if( fwrite( &pkh, n, 1, f_cap_out ) != 1 ) {
                     perror( "fwrite failed" );
+                    free(arp);
                     return( 1 );
                 }
 
@@ -3165,6 +3175,7 @@
 
                 if( fwrite( h80211, n, 1, f_cap_out ) != 1 ) {
                     perror( "fwrite failed" );
+                    free(arp);
                     return( 1 );
                 }
 
@@ -3173,6 +3184,7 @@
         }
     }
 
+    free(arp);
     return( 0 );
 }
 
