Index: src/besside-ng.c
===================================================================
--- src/besside-ng.c	(revision 2255)
+++ src/besside-ng.c	(working copy)
@@ -50,6 +50,7 @@
 #include <stdarg.h>
 #include <errno.h>
 #include <netdb.h>
+#include <unistd.h>
 
 #include "aircrack-ng.h"
 #include "version.h"
Index: src/Makefile
===================================================================
--- src/Makefile	(revision 2255)
+++ src/Makefile	(working copy)
@@ -47,25 +47,25 @@
 
 OBJS_AD		= airdecap-ng.o crypto.o common.o
 OBJS_PF		= packetforge-ng.o common.o crypto.o
-OBJS_AR		= aireplay-ng.o common.o crypto.o
-OBJS_ADU	= airodump-ng.o common.o crypto.o uniqueiv.o
+OBJS_AR		= aireplay-ng.o common.o crypto.o osdep/radiotap/radiotap-parser.o
+OBJS_ADU	= airodump-ng.o common.o crypto.o uniqueiv.o osdep/radiotap/radiotap-parser.o
 ifneq ($(OSNAME), Linux)
 	OBJS_ADU	+= osdep/common.o
 endif
-OBJS_AT		= airtun-ng.o common.o crypto.o
+OBJS_AT		= airtun-ng.o common.o crypto.o osdep/radiotap/radiotap-parser.o
 OBJS_IV		= ivstools.o common.o crypto.o uniqueiv.o
-OBJS_AS		= airserv-ng.o common.o
-OBJS_WS		= wesside-ng.o crypto.c common.o $(OBJS_PTW)
-OBJS_BS		= besside-ng.o crypto.c common.o $(OBJS_PTW)
+OBJS_AS		= airserv-ng.o common.o osdep/radiotap/radiotap-parser.o
+OBJS_WS		= wesside-ng.o crypto.c common.o osdep/radiotap/radiotap-parser.o $(OBJS_PTW)
+OBJS_BS		= besside-ng.o crypto.c common.o osdep/radiotap/radiotap-parser.o $(OBJS_PTW)
 OBJS_BC		= besside-ng-crawler.o
 OBJS_AL		= airolib-ng.o crypto.c common.o
-OBJS_ES		= easside-ng.o common.o
+OBJS_ES		= easside-ng.o common.o osdep/radiotap/radiotap-parser.o
 OBJS_BUDDY	= buddy-ng.o common.o
 OBJS_MI		= makeivs-ng.o common.o uniqueiv.o
-OBJS_AB		= airbase-ng.o common.o crypto.o
+OBJS_AB		= airbase-ng.o common.o crypto.o osdep/radiotap/radiotap-parser.o
 OBJS_AU		= airdecloak-ng.o common.o osdep/radiotap/radiotap-parser.o
-OBJS_TT		= tkiptun-ng.o common.o crypto.o
-OBJS_WC		= wpaclean.o
+OBJS_TT		= tkiptun-ng.o common.o crypto.o osdep/radiotap/radiotap-parser.o
+OBJS_WC		= wpaclean.o osdep/radiotap/radiotap-parser.o
 
 OSD		= osdep
 LIBS		:= -L$(OSD) -l$(OSD) $(LIBS)
