Index: aircrack-ng.c
===================================================================
--- aircrack-ng.c	(revision 2312)
+++ aircrack-ng.c	(working copy)
@@ -4589,12 +4589,12 @@
 
 int next_key( char **key, int keysize )
 {
-	char *tmp;
+	char *tmp, *tmpref;
 	int i, rtn;
 	unsigned int dec;
 	char *hex;
 
-	tmp = (char*) malloc(1024);
+	tmpref = tmp = (char*) malloc(1024);
 
 	while(1)
 	{
@@ -4604,7 +4604,7 @@
 		{
 			pthread_mutex_unlock( &mx_dic );
 			//printf( "\nPassphrase not in dictionary \n" );
-			free(tmp);
+			free(tmpref);
 			tmp = NULL;
 			return( FAILURE );
 		}
@@ -4623,7 +4623,7 @@
 //				printf( "\nPassphrase not in dictionary \"%s\" \n", opt.dicts[opt.nbdict] );
 				if(next_dict(opt.nbdict+1) != 0)
 				{
-					free(tmp);
+					free(tmpref);
 					tmp = NULL;
 					return( FAILURE );
 				}
@@ -4679,7 +4679,7 @@
 //				printf( "\nPassphrase not in dictionary \"%s\" \n", opt.dicts[opt.nbdict] );
 				if(next_dict(opt.nbdict+1) != 0)
 				{
-					free(tmp);
+					free(tmpref);
 					tmp = NULL;
 					return( FAILURE );
 				}
@@ -4702,9 +4702,7 @@
 		break;
 	}
 
-	free(tmp);
-	tmp = NULL;
-
+	free(tmpref);
 	return( SUCCESS );
 }
 